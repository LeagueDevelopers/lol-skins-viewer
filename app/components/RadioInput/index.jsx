import React, { PropTypes, PureComponent } from 'react';
import Option from 'components/RadioOption';
import { call } from 'utils';

import style from './index.scss';


/**
 * RadioInput Component
 *
 * @export
 * @class RadioInput
 * @extends {PureComponent}
 */
export default class RadioInput extends PureComponent {

  static propTypes = {
    children: PropTypes.node,
    disabled: PropTypes.bool,
    value: PropTypes.any,
    onClick: PropTypes.func,
    onChange: PropTypes.func,
    options: PropTypes.oneOfType([PropTypes.array, PropTypes.obj])
  }

  /**
   * Called when an option is selected. If the option
   * is different from what is currently selected,
   * onChange is called.
   * If the onClick prop was provided to the component,
   * it is called every time an option is clicked.
   *
   * @memberOf RadioInput
   * @param {Object} option the which was clicked
   */
  handleSelect = nextValue => {
    const { value, onChange, onClick } = this.props;
    call(onClick, nextValue);
    if (nextValue !== value) {
      call(onChange, nextValue);
    }
  }

  render () {
    const { children, options, disabled, value } = this.props;
    const shouldRenderOptions = options && options.map && options.map.call;
    return (
      <div className={style.radioInput}>
        {shouldRenderOptions && options.map(o =>
          <Option
            key={o.label}
            checked={o.value === value}
            disabled={disabled || o.disabled || false}
            onClick={() => this.handleSelect(o.value)}
            label={o.label}
            value={o.value}
          />)}
        {children}
      </div>
    );
  }
}
