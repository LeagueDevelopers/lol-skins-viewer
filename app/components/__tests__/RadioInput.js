import React from 'react';
import { shallow } from 'enzyme';

import RadioInput from '../RadioInput';

const options = [
  { value: 0, label: 'test0' },
  { value: 1, label: 'test1' },
  { value: 2, label: 'test2' },
  { value: 5, label: 'test3' },
];

describe.only('<RadioInput />', () => {
  it('should render correctly', () => {
    const props = {
      handleChange: jest.fn(),
      handleClick: jest.fn(),
      selectedOption: options[1],
      value: options[1].value,
    };
    const wrapper = shallow(<RadioInput options={options} {...props}>Hello</RadioInput>);
    expect(wrapper.get(0)).toMatchSnapshot();
  });

  it('should render as many children as the options it is provided', () => {
    const wrapper = shallow(<RadioInput options={options} />);
    expect(wrapper.children().length).toBe(options.length);
  });

  it('should render no children if no options are provided', () => {
    const wrapper = shallow(<RadioInput options={[]} />);
    expect(wrapper.children().length).toBe(0);
  });

  it('should render no children if options is undefined or otherwise lacks a map method', () => {
    const wrapper = shallow(<RadioInput />);
    expect(wrapper.children().length).toBe(0);
  });

  it('should render the correct label for each child', () => {
    const wrapper = shallow(<RadioInput options={options} />);
    wrapper.children().forEach((child, i) => {
      expect(child.props().label).toBe(options[i].label);
    });
  });

  it('should pass the disabled prop to its options', () => {
    const wrapper = shallow(<RadioInput options={options} disabled />);
    wrapper.children().forEach(child => {
      expect(child.props().disabled).toBe(true);
    });
  });

  it('should mark the selected option, using strict equality against the value prop', () => {
    const selectedOption = options[1];
    const value = selectedOption.value;
    const wrapper = shallow(<RadioInput options={options} value={value} />);
    const selected = wrapper.childAt(1);
    const unselectedNodes = wrapper.children()
      .map(child => child.props())
      .filter(child => !child.checked);
    expect(selected.props().checked).toBe(true);
    expect(unselectedNodes).toHaveLength(options.length - 1);
  });

  it('should call onChange with the new value if the selection changes', () => {
    const handleChange = jest.fn();
    const selectedOption = options[1];
    const value = selectedOption.value;

    const wrapper = shallow(<RadioInput options={options} value={value} onChange={handleChange} />);

    handleChange.mockImplementation(v => wrapper.setProps({ value: v }));

    const nextSelect = wrapper.childAt(3);

    nextSelect.simulate('click');
    expect(handleChange).toHaveBeenCalledWith(nextSelect.props().value);

    nextSelect.simulate('click');
    expect(handleChange).toHaveBeenCalledTimes(1);
  });

  it('should not call onChange with the new value if the selection does not change', () => {
    const handleChange = jest.fn();
    const selectedOption = options[1];
    const value = selectedOption.value;

    const wrapper = shallow(<RadioInput options={options} value={value} onChange={handleChange} />);

    handleChange.mockImplementation(v => wrapper.setProps({ value: v }));

    const nextSelect = wrapper.childAt(1);

    nextSelect.simulate('click');
    expect(handleChange).not.toHaveBeenCalled();
  });

  it('should call onClick even if the selection does not change', () => {
    const handleClick = jest.fn();
    const selectedOption = options[1];
    const value = selectedOption.value;

    const wrapper = shallow(<RadioInput options={options} value={value} onClick={handleClick} />);

    handleClick.mockImplementation(v => wrapper.setProps({ value: v }));

    const initialSelected = wrapper.childAt(1);
    expect(initialSelected.props().checked).toBe(true);

    const nextSelect = wrapper.childAt(3);

    nextSelect.simulate('click');
    expect(handleClick).toHaveBeenCalledWith(nextSelect.props().value);

    nextSelect.simulate('click');
    expect(handleClick).toHaveBeenCalledWith(nextSelect.props().value);
    expect(handleClick).toHaveBeenCalledTimes(2);
  });

  it('should fail gracefully if neither onClick or onChange are defined', () => {
    const selectedOption = options[1];
    const value = selectedOption.value;

    const wrapper = shallow(<RadioInput options={options} value={value} />);

    const nextSelect = wrapper.childAt(3);

    nextSelect.simulate('click');
  });
});
