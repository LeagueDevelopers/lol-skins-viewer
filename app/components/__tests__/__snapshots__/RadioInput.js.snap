exports[`<RadioInput /> should render correctly 1`] = `
<div
  className="radioInput">
  <RadioOption
    checked={false}
    disabled={false}
    label="test0"
    onClick={[Function]}
    value={0} />
  <RadioOption
    checked={true}
    disabled={false}
    label="test1"
    onClick={[Function]}
    value={1} />
  <RadioOption
    checked={false}
    disabled={false}
    label="test2"
    onClick={[Function]}
    value={2} />
  <RadioOption
    checked={false}
    disabled={false}
    label="test3"
    onClick={[Function]}
    value={5} />
  Hello
</div>
`;
